docs/specification/auth/oauth2.md

```diff
@@ -0,0 +1,285 @@
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "initialize",
+  "params": {
+    "protocolVersion": "...",
+    "capabilities": {
+      "auth": {
+        "oauth2": true
+      }
+    },
+    "headers": {
+      "Authorization": "Bearer access_token"
+    }
+  }
+}
+```
+
+### Invalid Access Token Flow
+
+If the access token is invalid, the server **MUST** respond with:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "error": {
+    "code": -32002,
+    "message": "Descriptive OAuth 2.0 issue (invalid, expired, etc.)",
+    "data": {
+      "authRequest": {
+        "authorizationUrl": "<scheme>://<domain>/oauth2",
+        "client_id": "client_id",
+        "scope": "scope",
+        "responseType": "code",
+        "state": "state"
+      }
+    }
+  }
+}
+```
+
+### Refresh Token Flow
+
+When an access token expires, the client **MUST** request a new one using the refresh token:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "auth/oauth2/token",
+  "params": {
+    "refresh_token": "refresh_token",
+    "grant_type": "refresh_token"
+  }
+}
+```
+
+Server response:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "accessToken": "access_token",
+    "refreshToken": "refresh_token",
+    "expiresIn": 3600,
+    "scope": "scope"
+  }
+}
+```
+
+### Revoke Token Flow
+
+To revoke a token:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "auth/oauth/revoke",
+  "params": {
+    "token": "access_or_refresh_token"
+  }
+}
+```
+
+## Error Handling
+
+Servers **SHOULD** return specific error codes for common OAuth 2.0 failure cases:
+
+- Invalid token: -32002
+- Expired token: -32002
+- Authorization required: -32001
+
+Example error response:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "error": {
+    "code": -32002,
+    "message": "Invalid OAuth 2.0 token",
+    "data": {
+      "authRequest": {
+        "authorizationUrl": "<scheme>://<domain>/oauth2",
+        "client_id": "client_id",
+        "scope": "scope",
+        "responseType": "code",
+        "state": "state"
+      }
+    }
+  }
+}
+```
+
+## Implementation Requirements
+
+### Client Requirements
+
+Clients **MUST**:
+1. Support the OAuth 2.0 authorization code flow
+2. Securely store access and refresh tokens
+3. Include valid tokens in request headers
+4. Implement token refresh logic
+5. Handle token revocation
+
+### Server Requirements
+
+Servers **MUST**:
+1. Validate all tokens
+2. Provide clear error messages for authentication failures
+3. Support token refresh
+4. Implement token revocation
+5. Maintain secure token storage
+
+## Security Considerations
+
+1. Clients **MUST**:
+   - Securely store tokens
+   - Validate state parameters
+   - Implement proper token refresh logic
+   - Handle token revocation
+
+2. Servers **MUST**:
+   - Validate all tokens
+   - Implement proper token expiration
+   - Support token revocation
+   - Maintain secure storage of client secrets
+   - Implement rate limiting for token endpoints
```

Comment:
This is a bit buried. Might be worth mentioning above as well. FWIW I agree this is should be a strong requirement on implementations.

Reply:
Moved up!